Lib/gzip.py

```diff
@@ -430,29 +430,42 @@ def _read_gzip_header(fp):
 
     if magic != b'\037\213':
         raise BadGzipFile('Not a gzipped file (%r)' % magic)
-
-    (method, flag, last_mtime) = struct.unpack("<BBIxx", _read_exact(fp, 8))
+    header_buffer = io.BytesIO()
+    header_buffer.write(magic)
+    base_header = _read_exact(fp, 8)
+    (method, flag, last_mtime) = struct.unpack("<BBIxx", base_header)
     if method != 8:
         raise BadGzipFile('Unknown compression method')
+    header_buffer.write(base_header)
 
     if flag & FEXTRA:
-        # Read & discard the extra field, if present
-        extra_len, = struct.unpack("<H", _read_exact(fp, 2))
-        _read_exact(fp, extra_len)
+        # Read the extra field, if present
+        extra_len_bytes = _read_exact(fp, 2)
+        header_buffer.write(extra_len_bytes)
+        extra_len, = struct.unpack("<H", extra_len_bytes)
+        header_buffer.write(_read_exact(fp, extra_len))
+
     if flag & FNAME:
         # Read and discard a null-terminated string containing the filename
         while True:
-            s = fp.read(1)
-            if not s or s==b'\000':
+            s = _read_exact(fp, 1)
+            header_buffer.write(s)
+            if s == b'\000':
                 break
     if flag & FCOMMENT:
         # Read and discard a null-terminated string containing a comment
         while True:
-            s = fp.read(1)
-            if not s or s==b'\000':
+            s = _read_exact(fp, 1)
+            header_buffer.write(s)
+            if s == b'\000':
                 break
     if flag & FHCRC:
-        _read_exact(fp, 2)     # Read & discard the 16-bit header CRC
+        # Read the 16-bit header CRC
+        header_crc, = struct.unpack("<H", _read_exact(fp, 2))
+        true_crc = zlib.crc32(header_buffer.getvalue()) & 0xFFFF
+        if header_crc != true_crc:
+            raise BadGzipFile(f"Corrupted gzip header. Checksums do not "
+                              f"match: {true_crc} != {header_crc}")
```

Comment:
Would not be more convenient to output the CRC values in hex? `:04x`

Reply:
Yes, this will make finding the actual header easier if this error is thrown. Done.